spark_log_parser/loaders/local_file.py

```diff
@@ -0,0 +1,28 @@
+from abc import ABC
+from pathlib import Path
+from urllib.parse import ParseResult, urlparse
+
+
+from spark_log_parser.loaders import AbstractFileDataLoader, AbstractLinesDataLoader, AbstractBlobDataLoader
+
+
+class AbstractLocalFileDataLoader(AbstractFileDataLoader, ABC):
+    """
+    Abstract class for loading files that are on a local disk. This assumes that all we need
+    is a path to the top-level file/directory from which we can extract data.
+    """
+    cache = False
+
+    def load_item(self, filepath: str | ParseResult):
+        parsed_url: ParseResult = filepath if isinstance(filepath, ParseResult) else urlparse(filepath)
+        path: Path = Path(parsed_url.path)
+
+        yield from self.extract(path)
+
+
+class LocalFileBlobDataLoader(AbstractLocalFileDataLoader, AbstractBlobDataLoader):
+    pass
+
+
+class LocalFileLinesDataLoader(AbstractLocalFileDataLoader, AbstractLinesDataLoader):
+    pass
```

Comment:
No, these (and the others) are fully implemented! The 2nd class, `AbstractBlobDataLoader`/`LinesDataLoader` provide the `read_file_stream` implementations, so these concrete classes are really just a composition of the appropriate abstract classes - https://github.com/synccomputingcode/spark_log_parser/pull/31/files/5042cdfa5f3b69695e8cc90218e5ae2d07abc543#diff-e6eadcc243cd6e0abd0b257ed701c0e4b5a24d8712a487a823c9f3a929f561c3R280

Reply:
Well that's cool! There may be an improvement that simplifies the inheritance hierarchy though. For example, `AbstractLinesDataLoader` could be replaced with a "mixin" that only provides `read_file_stream`. There's a similar example of using a "mixin" here: https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-is-it-useful

Reply:
Ah, yes, I've used mixins on interfaces in the JVM world. So the suggestion is to have some `LinesFileReaderMixin` that looks something like e.g. -

```python
class LinesFileReaderMixin:
    def read_file_stream(self, file: FileChunkStreamWrapper):
        yield from file.iter_lines()
```

And then this `LocalFileLinesDataLoader` could be something like - 

```python
class LocalFileLinesDataLoader(AbstractLocalFileDataLoader, LinesFileReaderMixin):
    ...
```
?

Reply:
Playing around with this a little bit, I'm not quite sure it will work... the reasons I can see for that are -

- `read_file_stream` is an `abstractmethod` on `AbstractFileDataLoader`. The Python runtime doesn't pick up that `LinesFileReaderMixin` (as written above) is actually providing an implementation, and errors out when trying to instantiate the class -
```python
FAILED tests/test_parse_local.py::test_simple_databricks_log - TypeError: Can't instantiate abstract class LocalFileLinesDataLoader with abstract method read_file_stream
ERROR tests/test_parse_local.py::test_simple_databricks_archive[parsed_files0] - TypeError: Can't instantiate abstract class LocalFileLinesDataLoader with abstract method read_fi...
ERROR tests/test_parse_local.py::test_simple_emr_log[parsed_files0] - TypeError: Can't instantiate abstract class LocalFileLinesDataLoader with abstract method read_file_stream
ERROR tests/test_parse_local.py::test_emr_missing_sql_events[parsed_files0] - TypeError: Can't instantiate abstract class LocalFileLinesDataLoader with abstract method read_file_...
ERROR tests/test_parse_local.py::test_databricks_rollover[parsed_files0] - TypeError: Can't instantiate abstract class LocalFileLinesDataLoader with abstract method read_file_stream
```
- To fix this, we can have the concrete class implement the abstract method and defer to the mixin, perhaps, but I can't actually figure out the correct `super()` incantation to call the `LinesFileReaderMixin` implementation of `read_file_stream`. I'm trying to do something like -
```python
class LocalFileLinesDataLoader(AbstractLocalFileDataLoader, LinesFileReaderMixin):

    def read_file_stream(self, file: FileChunkStreamWrapper) -> Iterator[bytes]:
        return super(LinesFileReaderMixin).read_file_stream(file)
```
but then I get errors like -
```
FAILED tests/test_parse_local.py::test_simple_databricks_log - AttributeError: 'super' object has no attribute 'read_file_stream'
```

Reply:
OK, I can get this inheritance hierarchy a bit "flatter" if we do something like -

```python
class AbstractFileReader(abc.ABC):
    @abc.abstractmethod
    def read_file_stream(self, file: FileChunkStreamWrapper) -> Iterator[bytes]:
        """

        """


class LinesFileReaderMixin(AbstractFileReader):
    """
    """

    def read_file_stream(self, file: FileChunkStreamWrapper):
        yield from file.iter_lines()


class AbstractFileDataLoader(AbstractFileReader, DataLoader, abc.ABC):
    ...
```

This moves `read_file_stream` up into an abstract class that our mixins and concretely implement, and that `AbstractFileDataLoader` can also inherit from in order to get the `self.read_file_stream()` calls to behave as expected.